resource/resource.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package resource defines the resource type and provides helpers to derive them as well
+// as the generic population through environment variables.
+package resource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+const (
+	envVarType = "OC_RESOURCE_TYPE"
+	envVarTags = "OC_RESOURCE_TAGS"
+)
+
+// Resource describes an entity about which data is exposed.
+type Resource struct {
+	Type string
+	Tags map[string]string
+}
+
+// EncodeTags encodes a tags to a string as provided via the OC_RESOURCE_TAGS environment variable.
+func EncodeTags(tags map[string]string) string {
+	sortedKeys := make([]string, 0, len(tags))
+	for k := range tags {
+		sortedKeys = append(sortedKeys, k)
+	}
+	sort.Strings(sortedKeys)
+
+	s := ""
+	for i, k := range sortedKeys {
+		if i > 0 {
+			s += ","
+		}
+		s += k + "=" + strconv.Quote(tags[k])
+	}
+	return s
+}
+
+var tagRegex = regexp.MustCompile(`\s*([a-zA-Z0-9-_./]+)=(?:(".*?")|([^\s="]+))\s*,`)
+
+// DecodeTags decodes serialized a tag map as used in the OC_RESOURCE_TAGS variable.
+// A list of tags of the form `<key1>=<value1>,<key2>=<value2>,...` is accepted.
+// Domain names and paths are accepted as tag keys. Values may be quoted or unquoted in general.
+// If a value contains whitespaces, =, or " characters, it must always be quoted.
+func DecodeTags(s string) (map[string]string, error) {
+	m := map[string]string{}
+	// Ensure a trailing comma, which allows us to keep the regex simpler
+	s = strings.TrimRight(strings.TrimSpace(s), ",") + ","
+
+	for len(s) > 0 {
+		match := tagRegex.FindStringSubmatch(s)
+		if len(match) == 0 {
+			return nil, fmt.Errorf("invalid tag formatting, remainder: %s", s)
+		}
+		v := match[2]
+		if v == "" {
+			v = match[3]
+		} else {
+			var err error
+			if v, err = strconv.Unquote(v); err != nil {
+				return nil, fmt.Errorf("invalid tag formatting, remainder: %s, err: %s", s, err)
+			}
+		}
+		m[match[1]] = v
+
+		s = s[len(match[0]):]
+	}
+	return m, nil
+}
+
+// FromEnv loads resource information from the OC_TYPE and OC_RESOURCE_TAGS environment variables.
+func FromEnv(context.Context) (*Resource, error) {
```

Comment:
I think it would be good documentation to have a type assertion that:
```go
var _ Detector = FromEnv
```

Reply:
I know something along those lines being used to statically assert interface implementation but doing this for documentation purposes (rather than just putting it into the doc strings) seems unusual in Go.